libs/x-content/src/main/java/org/elasticsearch/common/xcontent/XContentFactory.java

```diff
@@ -227,23 +229,49 @@ public static XContent xContent(byte[] data, int offset, int length) {
      */
     @Deprecated
     public static XContentType xContentType(InputStream si) throws IOException {
+        /*
+         * We need to guess the content type. To do this, we look for up to GUESS_HEADER_LENGTH bytes of non-whitespace characters on which
+         * we can attempt to do content detection. We repeatedly read GUESS_HEADER_LENGTH bytes from the stream until we find a
+         * non-whitespace character after which we take GUESS_HEADER_LENGTH bytes (or to the end of the stream). We do this in a way that
+         * does not modify the initial read position in the underlying input stream. This is why the input stream must support mark/reset
+         * and why we repeatedly mark the read position and reset.
+         */
         if (si.markSupported() == false) {
             throw new IllegalArgumentException("Cannot guess the xcontent type without mark/reset support on " + si.getClass());
         }
-        si.mark(GUESS_HEADER_LENGTH);
-        try {
-            final byte[] firstBytes = new byte[GUESS_HEADER_LENGTH];
-            int read = 0;
-            while (read < GUESS_HEADER_LENGTH) {
-                final int r = si.read(firstBytes, read, GUESS_HEADER_LENGTH - read);
-                if (r == -1) {
-                    break;
+        int iteration = 1;
+        while (true) {
+            si.mark(iteration * GUESS_HEADER_LENGTH);
```

Comment:
This is where is a bit uncomfortable with requiring the underlying stream to have an unbound support for mark. If I read things correctly we end up with a stream that just wraps a byte reference and we therefore don't care. If you're comfortable relying on that behavior, then I'm good but then I also think we don't need to allocate a growing size of `firstBytes` but rather read byte by byte until we find the first non-white space.

Reply:
I am comfortable with relying on that behavior. I pushed 86227baa5a7c80eec1cd3328b1c75d0d48c54402.